### ISSUE TITLE
Concurrency issue on on_completed

### ISSUE THREAD

**Issue by bittrance** [bittrance/rxruby#4]
Concurrency issue that occurs on `on_completed`. This may well be the issue that 4205fdd441003f4768650cb54a741826d0d615fd tried to address and that was reverted on the branch where this exception occurred:
```
rxruby-8c3dc69c5717/lib/rx/operators/multiple.rb:570:in `next': fiber called across threads (FiberError)
	from rxruby-8c3dc69c5717/lib/rx/operators/multiple.rb:570:in `block (3 levels) in concat'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/async_lock.rb:40:in `block in wait'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/async_lock.rb:27:in `loop'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/async_lock.rb:27:in `wait'
	from rxruby-8c3dc69c5717/lib/rx/operators/multiple.rb:561:in `block (2 levels) in concat'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/scheduler.rb:36:in `block in schedule_recursive'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/scheduler.rb:96:in `block in invoke_recursive'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/scheduler.rb:109:in `block (3 levels) in invoke_recursive'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/scheduled_item.rb:29:in `invoke'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/current_thread_scheduler.rb:67:in `run_trampoline'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/current_thread_scheduler.rb:40:in `schedule_relative_with_state'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/local_scheduler.rb:20:in `schedule_with_state'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/scheduler.rb:100:in `block (2 levels) in invoke_recursive'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/scheduler.rb:36:in `block (2 levels) in schedule_recursive'
	from rxruby-8c3dc69c5717/lib/rx/operators/multiple.rb:595:in `block (5 levels) in concat'
	from rxruby-8c3dc69c5717/lib/rx/core/observer.rb:106:in `on_completed'
	from rxruby-8c3dc69c5717/lib/rx/core/auto_detach_observer.rb:30:in `on_completed_core'
	from rxruby-8c3dc69c5717/lib/rx/core/observer.rb:106:in `on_completed'
	from rxruby-8c3dc69c5717/lib/rx/core/auto_detach_observer.rb:30:in `on_completed_core'
	from rxruby-8c3dc69c5717/lib/rx/core/observer.rb:106:in `on_completed'
	from project/lib/from_s3.rb:51:in `block in extract_from_file'
	from rxruby-8c3dc69c5717/lib/rx/operators/creation.rb:19:in `block in create'
	from rxruby-8c3dc69c5717/lib/rx/core/observable.rb:100:in `subscribe_core'
	from rxruby-8c3dc69c5717/lib/rx/core/observable.rb:46:in `_subscribe'
	from rxruby-8c3dc69c5717/lib/rx/core/observable.rb:23:in `subscribe'
	from rxruby-8c3dc69c5717/lib/rx/operators/synchronization.rb:24:in `block (2 levels) in subscribe_on'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/scheduler.rb:85:in `invoke'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/default_scheduler.rb:26:in `call'
	from rxruby-8c3dc69c5717/lib/rx/concurrency/default_scheduler.rb:26:in `block in schedule_with_state'
```
4205fdd441003f4768650cb54a741826d0d615fd introduced a trivial deadlock on unsubscribe, breaking test `test_unsubscribe_through_autodetach_observer` in https://github.com/bittrance/rxruby/blob/master/test/rx/subjects/test_subject.rb.

**Comment by bittrance** [comment]
Closer study of this issue indicates that the error originates from Ruby's enumerators, which are known not to be threadsafe. The question is if we can ensure that we have only one thread or fail properly otherwise, on the assumption that no enumerables in Ruby are threadsafe?

**Comment by bittrance** [comment]
Assuming this issue was a duplicate of #10 now closed as of release 0.1.0.
